# An Introduction

Performing any in-depth analysis is challenging. So over the years, our team has introduced different teaching styles to smoothen the learning curve. This series of notebooks will help guide your team through the fundamentals of data science and will show you a prime example of a winning analysis. We highly encourage everyone to read through the example and complete the workshops.

This year, we are going with a Wikipedia-style tutorial series where our tutorials for the week will be split into two different notebooks; one for the example analysis, and another for the workshops. Throughout the example, hyperlinks will connect the two workshops and will allow you to jump from notebook to notebook. These hyperlinks serve just like the ones on Wikipedia, they help round your knowledge if you come across an unfamiliar concept in the examples. We found that students learned best by supporting new concepts explained in the workshops with an actual example in an analysis. 

Nonetheless, go through the material in whatever order you like! Don't feel scared to complete the workshops first then view the examples. Remember to ask for help on Slack if we didn't do a good enough job at explaining!

Best wishes,<br>
The STEM Fellowship Data Science Team

# Workshop 1 - Data Handling
Welcome to the first workshop! This week we'll cover data handling and touch on the process data scientists use to convey a strong analysis.

There are three concepts that will be covered in this workshop:<br>
1. Data Storage
2. Data Indexing (Selecting subsets of your data)<br>
3. Data Manipulation

Let's begin by importing Pandas, Python's data handling library.

In [1]:
import pandas as pd

# Data Storage

Information in programs is usually stored in arrays and variables. Although we can store data in these data structures, data scientists usually use another data structure called a data frame. Think of a data frame as an Excel table. Like tables, they have columns, rows, and a header. Unlike arrays and variables, the table-like structure allows us to standardize information in a neat manner. Let's make our first data frame!

In [2]:
a2DArray = [["Alice","dog","female"],
            ["Bob","cat","male"],
            ["Grace","cat","female"]]


myFirstDataframe = pd.DataFrame(a2DArray,columns=["name","pet_type","gender"])
# The following variable is the last variable that is returned in the cell
myFirstDataframe

name pet_type  gender
0  Alice      dog  female
1    Bob      cat    male
2  Grace      cat  female

A few things to note:
 - The `.columns` attribute in the `pd.DataFrame()` command specifies the headers for the data frame.
 - The numbers on the left-hand side of the data frame are the row indexes. Notice how they were automatically generated.
 - Jupyter automatically displays the data frame if the data frame is the <b>last variable to be returned</b> in the cell. (in fact, this works with all Python objects)
 - We pass in a 2D array to make the data frame because tables (and subsequently data frames) are basically 2D arrays.
 
Feel free to run Python's `help()` function to learn more about unfamiliar functions!
(The *Parameters* section usually helps me out the most.)

In [3]:
#help(pd.DataFrame)

Now that we have a better understanding of the data frame, let's import a dataset that describes the amount of rainfall that fell on Toronto during May 2017. We have downloaded the data from their [website](https://www.toronto.ca/city-government/data-research-maps/open-data/open-data-catalogue/#09dee024-b840-174f-7270-29c1a1773d14) and have placed it within the datasets folder in this repository. Feel free to read the provided README file to get a better grasp of the dataset.

Most datasets are in the form of a CSV file (comma separated values). If you find a dataset that came as a JSON, HDF5, or even a shapefile, don't hesitate to ask how to import these files on Slack!

In [4]:
# Note how I am navigating through the file system to find the csv file I want.
# The directory you start in is always in relation to the directory of the current notebook
path = "/Users/curtis/Desktop/2019_bdc_workshops/"
df = pd.read_csv(path + "Datasets/Workshops/2017_rainfall_data/rainfall201705.csv")
df

id    name                 date  rainfall
0       7677  RG_001  2017-05-01T00:00:00     0.000
1       7677  RG_001  2017-05-01T00:05:00     0.254
2       7677  RG_001  2017-05-01T00:10:00     1.016
3       7677  RG_001  2017-05-01T00:15:00     0.000
4       7677  RG_001  2017-05-01T00:20:00     0.000
5       7677  RG_001  2017-05-01T00:25:00     0.254
6       7677  RG_001  2017-05-01T00:30:00     0.762
7       7677  RG_001  2017-05-01T00:35:00     0.000
8       7677  RG_001  2017-05-01T00:40:00     0.254
9       7677  RG_001  2017-05-01T00:45:00     0.000
10      7677  RG_001  2017-05-01T00:50:00     0.762
11      7677  RG_001  2017-05-01T00:55:00     0.508
12      7677  RG_001  2017-05-01T01:00:00     0.000
13      7677  RG_001  2017-05-01T01:05:00     0.000
14      7677  RG_001  2017-05-01T01:10:00     0.000
15      7677  RG_001  2017-05-01T01:15:00     0.000
16      7677  RG_001  2017-05-01T01:20:00     0.000
17      7677  RG_001  2017-05-01T01:25:00     0.254
18      7677  RG_001  2017-05-01T01:30:00     0.000
19      7677  RG_001  2017-05-01T01:35:00     0.000
20      7677  RG_001  2017-05-01T01:40:00     0.000
21      7677  RG_001  2017-05-01T01:45:00     0.000
22      7677  RG_001  2017-05-01T01:50:00     0.254
23      7677  RG_001  2017-05-01T01:55:00     0.000
24      7677  RG_001  2017-05-01T02:00:00     0.000
25      7677  RG_001  2017-05-01T02:05:00     0.000
26      7677  RG_001  2017-05-01T02:10:00     0.000
27      7677  RG_001  2017-05-01T02:15:00     0.000
28      7677  RG_001  2017-05-01T02:20:00     0.000
29      7677  RG_001  2017-05-01T02:25:00     0.000
...      ...     ...                  ...       ...
380017  8018  RG_056  2017-05-30T21:35:00     1.270
380018  8018  RG_056  2017-05-30T21:40:00     0.250
380019  8018  RG_056  2017-05-30T21:45:00     0.760
380020  8018  RG_056  2017-05-30T21:50:00     0.760
380021  8018  RG_056  2017-05-30T21:55:00     0.250
380022  8018  RG_056  2017-05-30T22:00:00     0.000
380023  8018  RG_056  2017-05-30T22:05:00     0.000
380024  8018  RG_056  2017-05-30T22:10:00     0.000
380025  8018  RG_056  2017-05-30T22:15:00     0.000
380026  8018  RG_056  2017-05-30T22:20:00     0.000
380027  8018  RG_056  2017-05-30T22:25:00     0.000
380028  8018  RG_056  2017-05-30T22:30:00     0.000
380029  8018  RG_056  2017-05-30T22:35:00     0.000
380030  8018  RG_056  2017-05-30T22:40:00     0.000
380031  8018  RG_056  2017-05-30T22:45:00     0.000
380032  8018  RG_056  2017-05-30T22:50:00     0.000
380033  8018  RG_056  2017-05-30T22:55:00     0.000
380034  8018  RG_056  2017-05-30T23:00:00     0.000
380035  8018  RG_056  2017-05-30T23:05:00     0.000
380036  8018  RG_056  2017-05-30T23:10:00     0.000
380037  8018  RG_056  2017-05-30T23:15:00     0.000
380038  8018  RG_056  2017-05-30T23:20:00     0.000
380039  8018  RG_056  2017-05-30T23:25:00     0.250
380040  8018  RG_056  2017-05-30T23:30:00     0.000
380041  8018  RG_056  2017-05-30T23:35:00     0.250
380042  8018  RG_056  2017-05-30T23:40:00     0.000
380043  8018  RG_056  2017-05-30T23:45:00     0.250
380044  8018  RG_056  2017-05-30T23:50:00     0.000
380045  8018  RG_056  2017-05-30T23:55:00     0.000
380046  8018  RG_056  2017-05-31T00:00:00     0.000

[380047 rows x 4 columns]

Whoa, that's a lot of data. But what is data anyways?
<B>Data</b> are values that describe something. Data is usually gathered from sensors or events that happen on your phone/computer (like a tap).<br>
<br>
Wait... What is information then?<br><br>
<b>Information</b> is data that is useful. Data is not meaningful. Information is.<br>
As a data scientist, your job is to turn <i>sets of data</i> (like the one we just loaded) into information. <br><br>
We call this transformation:

# The Data Science Process
Whenever we're working with a dataset we usually go through five stages:

1. Data Inspection
2. Data Preparation (Normalization and Cleaning)
3. Feature Engineering (Gathering Insights)
4. Generate the models
5. Storytelling

# Data Inspection
Every good data science project begins with data inspection. When we inspect the data we want to make sense of raw, unorganized values. The point of inspecting the data is to gather context on what we're looking at so we can begin to ask ourselves questions to answer!
Some things we can glimpse into are the means of the columns, the unique values of categorical variables, counts, max, mins, and modes. This should spark some questions if you see an abnormally high number in a column.

To take a 'glimpse' of the data, we can use the function `head()`. By default, this will give us the first 5 rows of the dataset.


### View first n rows

In [5]:
# display first 10 rows (using n=10, default n is 5)
#df.head(10)
df.head()

id    name                 date  rainfall
0  7677  RG_001  2017-05-01T00:00:00     0.000
1  7677  RG_001  2017-05-01T00:05:00     0.254
2  7677  RG_001  2017-05-01T00:10:00     1.016
3  7677  RG_001  2017-05-01T00:15:00     0.000
4  7677  RG_001  2017-05-01T00:20:00     0.000

### View General Data
Much better, now we can save on screen estate when taking a look at our data frames.

Briefly:
 - The `id` column represents the station id that the data was gathered from.
 - The `name` column is the name of the station. "RG_001" represents "Rain Guage 001".
 - The `date` column represents the time the measurement was taken
 - The `rainfall` column represents the amount of rainfall the stations measured in mm over the time period.
 
If you look at the date column you'll notice that the measurements were taken every 5 minutes. This means that the rainfall column represents the amount of rain that fell over the past 5 minutes.


Let's dive a little deeper and use `describe()` to learn about the specifics of the dataset.

In [6]:
df.describe()

id       rainfall
count  380047.000000  375873.000000
mean     7717.234287       0.018821
std        84.255586       0.101155
min      7674.000000       0.000000
25%      7685.000000       0.000000
50%      7697.000000       0.000000
75%      7708.000000       0.000000
max      8049.000000       5.842000

You may notice that we did not get information about the name or the date. This is because the `describe()` function only gives us statistical summaries of numeric data. Don't worry, we'll go over other exploration techniques for categorical data later.

*Note: std means standard deviation and the numbers at the 25%, 50%, and 75% mark are [numeric percentiles](https://www.mathsisfun.com/data/percentiles.html) of the quartile ranges.*
We can also use the `sum()`, `mean()`, `min()`, `std()` and `unique()` on specific columns of the data frame to learn more about it.

*Note 2: Don't forget that a great analysis always places statistics at the forefront. Try to look up statistics libraries (such as [statsmodels](https://www.statsmodels.org/stable/index.html)) to learn more about your data!*

### View the headers of the data frame
For larger datasets, you'll often come across many different columns. In such cases, it will be hard to know what columns you are looking at. Fortunately, the `.columns` attribute helps you see all of the headers of your data frame.

In [7]:
df.columns

Index(['id', 'name', 'date', 'rainfall'], dtype='object')

### View a certain range of rows of the data frame (Indexing)
When inspecting datasets we might be interested in viewing different sections of the data frame. The process of selecting different subsets of our data is called indexing.

In [8]:
# display rows 5 to 10
df[5:10]

id    name                 date  rainfall
5  7677  RG_001  2017-05-01T00:25:00     0.254
6  7677  RG_001  2017-05-01T00:30:00     0.762
7  7677  RG_001  2017-05-01T00:35:00     0.000
8  7677  RG_001  2017-05-01T00:40:00     0.254
9  7677  RG_001  2017-05-01T00:45:00     0.000

### View a column of the data frame¶
Great job! You may notice that the left-hand side of the data frame shows the index. Now let's take a moment to read some of the headers and see what sparks your interest. For me, the column `name` seems interesting. I want to know how many different station names there are in the dataset. Let's start by selecting this specific column from the data frame.

In [9]:
df['name']
#pro tip: try running df['name'][:5] to only see the first 5 rows

0         RG_001
1         RG_001
2         RG_001
3         RG_001
4         RG_001
5         RG_001
6         RG_001
7         RG_001
8         RG_001
9         RG_001
10        RG_001
11        RG_001
12        RG_001
13        RG_001
14        RG_001
15        RG_001
16        RG_001
17        RG_001
18        RG_001
19        RG_001
20        RG_001
21        RG_001
22        RG_001
23        RG_001
24        RG_001
25        RG_001
26        RG_001
27        RG_001
28        RG_001
29        RG_001
           ...  
380017    RG_056
380018    RG_056
380019    RG_056
380020    RG_056
380021    RG_056
380022    RG_056
380023    RG_056
380024    RG_056
380025    RG_056
380026    RG_056
380027    RG_056
380028    RG_056
380029    RG_056
380030    RG_056
380031    RG_056
380032    RG_056
380033    RG_056
380034    RG_056
380035    RG_056
380036    RG_056
380037    RG_056
380038    RG_056
380039    RG_056
380040    RG_056
380041    RG_056
380042    RG_056
380043    RG_056
380044    RG_0

Every time we do a selection command on parts of the data frame we are **<u>not changing the original data frame<u>.** Instead, we are returning the column in the cell and Jupyter displays the last returned value in the cell. In this case, the last returned value is the `name` column of the data frame.<br><br>
*Note: This column is not a data frame in itself. It is actually a Pandas `series` object. Think of a series object as a 1D array. We will explain the differences between a series and a data frame later on in this notebook but keep this distinction in mind when you receive an error for accidentally running a data frame operation on a series.*

### View unique values in column
Now that we have the column selected we will use Pandas' `unique()` function to print out all the unique station names.

*Fun fact: the `unique` command only works on a `series` object.*

In [10]:
df['name'].unique()

array(['RG_001', 'RG_002', 'RG_003', 'RG_004', 'RG_006', 'RG_007',
       'RG_012', 'RG_013', 'RG_014', 'RG_015', 'RG_016', 'RG_017',
       'RG_018', 'RG_019', 'RG_020', 'RG_021', 'RG_022', 'RG_023',
       'RG_024', 'RG_025', 'RG_027', 'RG_028', 'RG_030', 'RG_031',
       'RG_033', 'RG_034', 'RG_035', 'RG_036', 'RG_037', 'RG_038',
       'RG_039', 'RG_040', 'RG_041', 'RG_042', 'RG_044', 'RG_045',
       'RG_046', 'RG_047', 'RG_048', 'RG_049', 'RG_051', 'RG_052',
       'RG_054', 'RG_055', 'RG_056'], dtype=object)

### Counting the number of unique columns
If you ever want to count the number of unique items in a column just use Python's `len()` function.

In [11]:
len(df['name'].unique())

45

If we wanted to take a shortcut by looking at the last station name (`RG_056`) we would erroneously assume that there are 56 different stations. By using the `len()` function we avoided this mistake.<br><br>

One possible explanation for the missing identification numbers is that some of the stations were retired or are still being constructed. It's always important to break down your assumptions when working with data!

For kicks, I want to check if there are the same number of unique `id`s as `names`s

In [12]:
len(df["id"].unique())

45

### Removing a column
Sometimes you might want to remove columns because they don't add extra value to your analysis. In this case, I might want to remove the `id` column because these stations are already identified by `name`. Since the number of unique `id`s is the same as the number of unique `name`s the `id` column is basically a repeat of the `name` column (I chose to remove the `id` column because I might want to use the station's name and look for it on a map). Other reasons to remove columns are when they are completely homogenous, giving you no information about the data (unless its homogeneity is unusual).<br><br> Anyhow, let's remove this column.

In [13]:
# axis = 0 refers to the rows, axis = 1 refers to the columns
# errors="ignore" will stop this cell from complaining when we have already dropped the column
df = df.drop(['id'],axis=1,errors="ignore")

You may have noticed that we assigned `df` to the returned value of `drop()` function. For all Pandas functions, be it data frames or series (the data type that represents a data frame column), we have to assign the original value back to the modified value because **<u>all of these functions do not modify the original data frame, they only return the modified data.</u>**

### View two columns of the data frame
You can also select multiple columns at once if you want to narrow your focus. Let's select the `date` and `rainfall` columns of the data frame.

In [14]:
df[['date','rainfall']]

date  rainfall
0       2017-05-01T00:00:00     0.000
1       2017-05-01T00:05:00     0.254
2       2017-05-01T00:10:00     1.016
3       2017-05-01T00:15:00     0.000
4       2017-05-01T00:20:00     0.000
5       2017-05-01T00:25:00     0.254
6       2017-05-01T00:30:00     0.762
7       2017-05-01T00:35:00     0.000
8       2017-05-01T00:40:00     0.254
9       2017-05-01T00:45:00     0.000
10      2017-05-01T00:50:00     0.762
11      2017-05-01T00:55:00     0.508
12      2017-05-01T01:00:00     0.000
13      2017-05-01T01:05:00     0.000
14      2017-05-01T01:10:00     0.000
15      2017-05-01T01:15:00     0.000
16      2017-05-01T01:20:00     0.000
17      2017-05-01T01:25:00     0.254
18      2017-05-01T01:30:00     0.000
19      2017-05-01T01:35:00     0.000
20      2017-05-01T01:40:00     0.000
21      2017-05-01T01:45:00     0.000
22      2017-05-01T01:50:00     0.254
23      2017-05-01T01:55:00     0.000
24      2017-05-01T02:00:00     0.000
25      2017-05-01T02:05:00     0.000
26      2017-05-01T02:10:00     0.000
27      2017-05-01T02:15:00     0.000
28      2017-05-01T02:20:00     0.000
29      2017-05-01T02:25:00     0.000
...                     ...       ...
380017  2017-05-30T21:35:00     1.270
380018  2017-05-30T21:40:00     0.250
380019  2017-05-30T21:45:00     0.760
380020  2017-05-30T21:50:00     0.760
380021  2017-05-30T21:55:00     0.250
380022  2017-05-30T22:00:00     0.000
380023  2017-05-30T22:05:00     0.000
380024  2017-05-30T22:10:00     0.000
380025  2017-05-30T22:15:00     0.000
380026  2017-05-30T22:20:00     0.000
380027  2017-05-30T22:25:00     0.000
380028  2017-05-30T22:30:00     0.000
380029  2017-05-30T22:35:00     0.000
380030  2017-05-30T22:40:00     0.000
380031  2017-05-30T22:45:00     0.000
380032  2017-05-30T22:50:00     0.000
380033  2017-05-30T22:55:00     0.000
380034  2017-05-30T23:00:00     0.000
380035  2017-05-30T23:05:00     0.000
380036  2017-05-30T23:10:00     0.000
380037  2017-05-30T23:15:00     0.000
380038  2017-05-30T23:20:00     0.000
380039  2017-05-30T23:25:00     0.250
380040  2017-05-30T23:30:00     0.000
380041  2017-05-30T23:35:00     0.250
380042  2017-05-30T23:40:00     0.000
380043  2017-05-30T23:45:00     0.250
380044  2017-05-30T23:50:00     0.000
380045  2017-05-30T23:55:00     0.000
380046  2017-05-31T00:00:00     0.000

[380047 rows x 2 columns]

*Note: we are not changing the value of the data frame because we did not assign the returned value back into `df`.*

### View the nth index
This is pretty useful for keeping tabs on rows that stand out.

In [15]:
df.iloc[3]

name                     RG_001
date        2017-05-01T00:15:00
rainfall                      0
Name: 3, dtype: object

### View the first ten rows by index and the first two columns
Now let's try to select the data frame by its row and by its column at the same time! Notice how `name` and `date` are the first two indexes of the data frame.

In [16]:
df.iloc[0:10,0:2]

name                 date
0  RG_001  2017-05-01T00:00:00
1  RG_001  2017-05-01T00:05:00
2  RG_001  2017-05-01T00:10:00
3  RG_001  2017-05-01T00:15:00
4  RG_001  2017-05-01T00:20:00
5  RG_001  2017-05-01T00:25:00
6  RG_001  2017-05-01T00:30:00
7  RG_001  2017-05-01T00:35:00
8  RG_001  2017-05-01T00:40:00
9  RG_001  2017-05-01T00:45:00

### Merging Datasets

Viewing your data is always important to make sure that you are manipulating your data correctly. What we have looked at so far is only half the dataset! There are also a set of files that describe the locations of the stations that took these measurements. Let's use our newfound inspection tools and merge the second half of the dataset with this one.<br><br>
We'll start by importing the second dataset.

In [17]:
df2 = pd.read_csv(path + "Datasets/Workshops/2017_rainfall_data/sites201706.csv")
print("length of df2: %s"% len(df2))
df2.head()

length of df2: 46


id    name  longitude   latitude
0  7677  RG_001 -79.478112  43.647680
1  7678  RG_002 -79.443616  43.651196
2  7679  RG_003 -79.405094  43.656616
3  7680  RG_004 -79.402833  43.678343
4  7681  RG_006 -79.375101  43.661265

Our goal is to merge the `latitude` and `longitude` columns into the first data frame. 

Again we'll remove the `id` column because it is redundant.

In [18]:
df2 = df2.drop("id",axis=1,errors="ignore")
df2.head()

name  longitude   latitude
0  RG_001 -79.478112  43.647680
1  RG_002 -79.443616  43.651196
2  RG_003 -79.405094  43.656616
3  RG_004 -79.402833  43.678343
4  RG_006 -79.375101  43.661265

Now let's do the actual merging. In this case, for every station name in the first data frame, we will try to find a corresponding row in `df2` and attach the `latitude` and `longitude` columns into one large data frame. If you ever find merging data frames confusing take a look at the [official Pandas merging documentation](https://pandas.pydata.org/pandas-docs/stable/merging.html). They have great visuals to explain the different merging functions.<br><br>

For us, this is an [outer merge](https://pandas.pydata.org/pandas-docs/stable/merging.html#brief-primer-on-merge-methods-relational-algebra)

In [19]:
df3 = df.merge(df2, left_on='name', right_on='name', how='outer')
df3.head()

name                 date  rainfall  longitude  latitude
0  RG_001  2017-05-01T00:00:00     0.000 -79.478112  43.64768
1  RG_001  2017-05-01T00:05:00     0.254 -79.478112  43.64768
2  RG_001  2017-05-01T00:10:00     1.016 -79.478112  43.64768
3  RG_001  2017-05-01T00:15:00     0.000 -79.478112  43.64768
4  RG_001  2017-05-01T00:20:00     0.000 -79.478112  43.64768

Fantastic. Now we have a data frame that tells us the name of the station, when the recording happened, the amount of rain that fell, and the geographic coordinates of the reading. More importantly, you saw how selection functions like `head()` was integral to check our work.

### Data Filtering
Filtering data frames allow you to view the datasets in a certain manner. Filtering your data can help you derive great insights about your dataset.

Let's say we want to find how many rows in the above data frame has rainfall greater than 0.

In [20]:
rain_greater_than_zero_df = df3[df3['rainfall']>0]
rain_greater_than_zero_df.head()

name                 date  rainfall  longitude  latitude
1  RG_001  2017-05-01T00:05:00     0.254 -79.478112  43.64768
2  RG_001  2017-05-01T00:10:00     1.016 -79.478112  43.64768
5  RG_001  2017-05-01T00:25:00     0.254 -79.478112  43.64768
6  RG_001  2017-05-01T00:30:00     0.762 -79.478112  43.64768
8  RG_001  2017-05-01T00:40:00     0.254 -79.478112  43.64768

In [21]:
len(rain_greater_than_zero_df)

20586

Wow, 20586 seems really small. What percentage of all measurements actually detect rainfall?

In [22]:
len(rain_greater_than_zero_df) / len(df3)

0.05416684208310529

This measurement states that for the month of June, rain fell on Toronto 5% of the time. Even though this number feels surprisingly low, it is pretty accurate. In fact, it is above the average rate of rainfall that Toronto receives. We encourage you to see how this number compares to other months in the dataset.<br><br>
*bonus: Slack your reason for the abnormally high rainfall and tell us what happened to Toronto in May of 2017!*<br><br>
Anyway, Let's move on and try to select all the "rainfall" readings from the station "RG_016" which are greater than zero.

In [23]:
# Notice the AND (&) operator.
rainfall_and_RG_016_df = df3[ (df3['name']=="RG_016") & (df3['rainfall']>0) ]
rainfall_and_RG_016_df.head()

name                 date  rainfall  longitude   latitude
87502  RG_016  2017-05-07T08:55:00     0.254 -79.520366  43.608125
87954  RG_016  2017-05-08T22:35:00     0.254 -79.520366  43.608125
89589  RG_016  2017-05-14T14:55:00     0.254 -79.520366  43.608125
90116  RG_016  2017-05-16T10:50:00     0.254 -79.520366  43.608125
90782  RG_016  2017-05-18T18:20:00     0.254 -79.520366  43.608125

Let's also try to select the readings that are greater than 0.2 OR the readings that came from station "RG_018".

In [24]:
rainfall_greater_than_02_OR_RG_018_df = df3[ (df3['rainfall']>0.2) | (df3['name']=="RG_018") ]
rainfall_greater_than_02_OR_RG_018_df.head()

name                 date  rainfall  longitude  latitude
1  RG_001  2017-05-01T00:05:00     0.254 -79.478112  43.64768
2  RG_001  2017-05-01T00:10:00     1.016 -79.478112  43.64768
5  RG_001  2017-05-01T00:25:00     0.254 -79.478112  43.64768
6  RG_001  2017-05-01T00:30:00     0.762 -79.478112  43.64768
8  RG_001  2017-05-01T00:40:00     0.254 -79.478112  43.64768

Selecting different sections of a dataset is not only useful for data inspection but it is also applied to:


# Data Preparation (Normalization and Cleaning)

Now that we have a good sense of the data, we should use our newfound selection abilities to prepare the data. <br><br>
What do you mean by "preparing the data" and why do we need to do it?<br><br>
Data preparation is another step to turn raw data into useful information. When we prepare the data we are making it simple for visualizations and eventually modelling. This involves converting variables into the right time format, units, understandable text, and even altering the ranges that values fall on. When we are done, the data is cleaned because it becomes easier for people to parse and is processable by algorithms.<br><br>
We'll start off our data preparation by removing empty values within our dataset.

In [25]:
print("length of df3 before removing bad data: %d"%len(df3))
df3 = df3.dropna()
print("length of df3 after removing bad data: %d"%len(df3))

length of df3 before removing bad data: 380048
length of df3 after removing bad data: 375873


By removing the empty values we lost 10 thousand rows. If your data set is much smaller please be cognizant of the amount of data you are removing. You might also choose to keep the empty data if it leads to new insights. Have good judgement and try to keep statistical significance in the back of your mind. In our case, 10 thousand rows is a drop in the bucket when we have 347 thousand rows left.

### View the data types of all columns

Another thing to keep in mind when preparing your data is the type of the columns in your data frame. You want numeric columns to be a numeric data type so you can run calculations on them. It's always important to convert columns into their appropriate data types.

To check the data types we call the `.dtypes` attribute of the data frame.

In [26]:
df3.dtypes

name          object
date          object
rainfall     float64
longitude    float64
latitude     float64
dtype: object

For us, we were really lucky because most of our columns are the correct data type. If you're unsure about the type of each column, remember that:

- **Categorical Variables** are objects
- **Numerical Variables** are numeric types

This is especially tricky when a categorical variable is given to you in the form of a number. Once you convert these columns into the dtype object, you can use categorical functions such as `unique()`.

### Converting the data types of a column
The one column that we should convert is the `date` column. Right now it is a string but we should convert it to a Python `datetime` object to leverage the suite of Python tools that selects time-specific data.<br><br>
Let's use the Pandas `to_datetime()` [function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.to_datetime.html#pandas.to_datetime).

If you get stuck, check out the [official Pandas documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.astype.html) for converting dtypes.

In [27]:
df3["date"] = pd.to_datetime(df3["date"])
df3.head()

name                date  rainfall  longitude  latitude
0  RG_001 2017-05-01 00:00:00     0.000 -79.478112  43.64768
1  RG_001 2017-05-01 00:05:00     0.254 -79.478112  43.64768
2  RG_001 2017-05-01 00:10:00     1.016 -79.478112  43.64768
3  RG_001 2017-05-01 00:15:00     0.000 -79.478112  43.64768
4  RG_001 2017-05-01 00:20:00     0.000 -79.478112  43.64768

Notice how the dtype changed:

In [28]:
df3.dtypes

name                 object
date         datetime64[ns]
rainfall            float64
longitude           float64
latitude            float64
dtype: object

The data cleaning process will differ for every data set. You might have to modify random strings from certain columns. You may also have to change outliers that are the result of input error. Whatever the cause, you will have to learn how to change and modify columns; an integral step which precedes:

# Feature Engineering (Gathering Insights)

We're on the last topic for this week. If **information** is useful **data**, then **features** are useful **columns**. Often, the data we gather from datasets may not contain any useful meaning. <br><br>
**Feature Engineering** is the extraction of meaningful information out of the variables in our dataset.<br><br>
An example of feature engineering is determining the day of the week given a date. We are creating a new feature (the day of the week) given a variable (the date). Another example is extracting the surname out of a full name. This is considered feature engineering because surnames can identify people who are part of the same family in a dataset.

### Creating a new Column
Let's start by creating an empty column from the dataset.

In [29]:
#this will initialize a new column in df_rain to 0
df3["empty"] = 0

In [30]:
df3.head()

name                date  rainfall  longitude  latitude  empty
0  RG_001 2017-05-01 00:00:00     0.000 -79.478112  43.64768      0
1  RG_001 2017-05-01 00:05:00     0.254 -79.478112  43.64768      0
2  RG_001 2017-05-01 00:10:00     1.016 -79.478112  43.64768      0
3  RG_001 2017-05-01 00:15:00     0.000 -79.478112  43.64768      0
4  RG_001 2017-05-01 00:20:00     0.000 -79.478112  43.64768      0

### Manipulating columns using basic arithmetic operators

One common way to make meaningful features is by using an arithmetic operator on multiple columns of the data frame.

In [31]:
#arithmetic operators
df3["lat_minus_long"] = df3["latitude"] - df3["longitude"]
df3.head()

name                date  rainfall  longitude  latitude  empty  \
0  RG_001 2017-05-01 00:00:00     0.000 -79.478112  43.64768      0   
1  RG_001 2017-05-01 00:05:00     0.254 -79.478112  43.64768      0   
2  RG_001 2017-05-01 00:10:00     1.016 -79.478112  43.64768      0   
3  RG_001 2017-05-01 00:15:00     0.000 -79.478112  43.64768      0   
4  RG_001 2017-05-01 00:20:00     0.000 -79.478112  43.64768      0   

   lat_minus_long  
0      123.125792  
1      123.125792  
2      123.125792  
3      123.125792  
4      123.125792

Let's clean up our data frame again.

In [32]:
df3 = df3.drop(["empty","lat_minus_long"],axis=1,errors="ignore")
df3.head()

name                date  rainfall  longitude  latitude
0  RG_001 2017-05-01 00:00:00     0.000 -79.478112  43.64768
1  RG_001 2017-05-01 00:05:00     0.254 -79.478112  43.64768
2  RG_001 2017-05-01 00:10:00     1.016 -79.478112  43.64768
3  RG_001 2017-05-01 00:15:00     0.000 -79.478112  43.64768
4  RG_001 2017-05-01 00:20:00     0.000 -79.478112  43.64768

A more practical feature would be the rainfall measured in centimetres. For that, we'll use the multiplication operator.

In [33]:
df3["rainfall_in_cm"] = df3["rainfall"] * 10
df3.head()

name                date  rainfall  longitude  latitude  rainfall_in_cm
0  RG_001 2017-05-01 00:00:00     0.000 -79.478112  43.64768            0.00
1  RG_001 2017-05-01 00:05:00     0.254 -79.478112  43.64768            2.54
2  RG_001 2017-05-01 00:10:00     1.016 -79.478112  43.64768           10.16
3  RG_001 2017-05-01 00:15:00     0.000 -79.478112  43.64768            0.00
4  RG_001 2017-05-01 00:20:00     0.000 -79.478112  43.64768            0.00

### Manipulating columns using native Python functions
Sometimes you want to use a Python function on an entire column. The most useful of these functions are usually string operators. This is best explained by an example so I will extract the id of the station name by removing the "RG_" in the `name` column. <br><br>
To do this, I will use a `lambda` function. In Python, we use lambda functions to quickly create a one-time use function without a variable name. By using the `map()` function in Pandas, we can apply this lambda function to every row of our dataset.<br><br>
The function I am using is a string manipulation method. This method returns all the characters of a string after the third character, thus removing the pesky "RG_" in the `name` column.

In [34]:
df3["name_without_RG_"] = df3["name"].map(lambda x: x[3:])
df3.head()

name                date  rainfall  longitude  latitude  rainfall_in_cm  \
0  RG_001 2017-05-01 00:00:00     0.000 -79.478112  43.64768            0.00   
1  RG_001 2017-05-01 00:05:00     0.254 -79.478112  43.64768            2.54   
2  RG_001 2017-05-01 00:10:00     1.016 -79.478112  43.64768           10.16   
3  RG_001 2017-05-01 00:15:00     0.000 -79.478112  43.64768            0.00   
4  RG_001 2017-05-01 00:20:00     0.000 -79.478112  43.64768            0.00   

  name_without_RG_  
0              001  
1              001  
2              001  
3              001  
4              001

### Manipulating columns using custom functions
In some cases, your manipulation might include complex statements like loops or if statements. When this occurs, `lambda` functions aren't flexible enough to manipulate the column, so you might want to generate a custom function. To demonstrate, let's try to separate the rainfall readings into two groups, readings above the center latitude of Toronto and readings below the center latitude. According to Google, Toronto is located at <br><br>43.6532° N, 79.3832° W <br>or<br> lat: 43.6532 lon: -79.3832<br><br>
To perform this separation we have to define a function that takes in the current latitude and returns 1 if the reading is above 43.6532.<br><br>
We will still use the `map()` and `lambda()` functions but this time we will pass in our own function.

In [35]:
def is_above_toronto_midpoint_latitude(reading):
    if reading > 43.6532:
        return 1
    else:
        return 0

df3["is_above_toronto_midpoint_latitude"] = df3["latitude"].map(lambda x: is_above_toronto_midpoint_latitude(x))
df3.head()

name                date  rainfall  longitude  latitude  rainfall_in_cm  \
0  RG_001 2017-05-01 00:00:00     0.000 -79.478112  43.64768            0.00   
1  RG_001 2017-05-01 00:05:00     0.254 -79.478112  43.64768            2.54   
2  RG_001 2017-05-01 00:10:00     1.016 -79.478112  43.64768           10.16   
3  RG_001 2017-05-01 00:15:00     0.000 -79.478112  43.64768            0.00   
4  RG_001 2017-05-01 00:20:00     0.000 -79.478112  43.64768            0.00   

  name_without_RG_  is_above_toronto_midpoint_latitude  
0              001                                   0  
1              001                                   0  
2              001                                   0  
3              001                                   0  
4              001                                   0

### Manipulating columns by comparing to previous rows in a data frame
Let's find the difference in rainfall. This feature might help us see how rainfall changes over time. We can use Panda's `shift()` function to compare the current row to the previous row:

In [36]:
df3["rainfall_delta"] = df3["rainfall"] - df3["rainfall"].shift(1)
df3.head()

name                date  rainfall  longitude  latitude  rainfall_in_cm  \
0  RG_001 2017-05-01 00:00:00     0.000 -79.478112  43.64768            0.00   
1  RG_001 2017-05-01 00:05:00     0.254 -79.478112  43.64768            2.54   
2  RG_001 2017-05-01 00:10:00     1.016 -79.478112  43.64768           10.16   
3  RG_001 2017-05-01 00:15:00     0.000 -79.478112  43.64768            0.00   
4  RG_001 2017-05-01 00:20:00     0.000 -79.478112  43.64768            0.00   

  name_without_RG_  is_above_toronto_midpoint_latitude  rainfall_delta  
0              001                                   0             NaN  
1              001                                   0           0.254  
2              001                                   0           0.762  
3              001                                   0          -1.016  
4              001                                   0           0.000

For the time being, this should be sufficient for you to generate useful data. If you want to perform more complex operations by using values from previous rows and values from the current column, you will have to loop through every row of the data frame by using the methods `apply()` or `itertuples()`.

## Saving Data frames

After manipulating our data frame, we should save our work by saving the data frame we made (so we don't have to process everything all over again everytime we close Jupyter). The `.pickle` file type can store Python data frames as a .pkl file. To save our data frame, run this command:

In [37]:
df3.to_pickle(path + "Saved_Dataframes/workshop1RainfallDF")
# Note how we're saving our pickle file into the saved_dataframes folder.
# Keeping all your data in one place will keep you organized.

# Storytelling
You've earned your first wings detective! The purpose of data science is to tell a story. We crunch numbers to discover insights and to reinforce decisions. So as you chip away at your project remember to lace together a storyline.<br><br>
You have a long way to go, but for now, congratulations on completing the first workshop! Don't forget to view the example analysis to reinforce what you've learned. You can now manipulate and clean, and, using what you have learned, no dataset will be too big!